content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+
+You can actually create your own version of `callPackage`, which comes in quite
+handy when you have large sets where the attributes to be built depend on each
+other.
+
+```nix
+# In the next examples I will not implement or show the "called" 
+# files, as I think they are not necessary to understand the point I
+# want to make.
+```
+
+Consider the following initial version:
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> { }; in
+rec {
+  a = pkgs.callPackage ./a.nix { };
+  b = pkgs.callPackage ./b.nix { inherit a; };
+  c = pkgs.callPackage ./c.nix { inherit b; };
+  d = pkgs.callPackage ./d.nix { };
+  e = pkgs.callPackage ./e.nix { inherit c d; };
+}
+```
+
+Here you have to remember passing required arguments that are not in nixpkgs'
+toplevel manually.
+
+This can become quite tedious quickly, especially the larger the set becomes.
+
+Therefore we can use `lib.callPackageWith` to create our own `callPackage` version.
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+  callPackage = lib.callPackageWith (pkgs // packages);
+  packages = {
+    a = callPackage ./a.nix { };
+    b = callPackage ./b.nix { };
+    c = callPackage ./c.nix { };
+    d = callPackage ./d.nix { };
+    e = callPackage ./e.nix { };
+  };
+in
+  packages
+```
+
+Our modified `callPackage` now will exactly "know" how to resolve the dependencies
+through the set defined by `pkgs // packages`.
+
+Nix' laziness does us a good favour here and makes this actually possible.
+
+## Summary
+
+So using `callPackage` doesn't only make your code uniform to what you see a lot
+in nixpkgs already, it also gives you some things for free:
+
+1. parametrized builds
+2. overrideable builds
+3. cleaner implementation of large interdepending package sets
+
+## Further reading
+
+There is also `callPackages` and `lib.callPackages` which do a pretty similar
+thing, though they expect that the return value is *not* a package, but a
+packageset.
+Each of the attributes in the returned set will then be overrideable as if you
+had called `callPackage` on that.
+
+```nix
+# callpackages.nix
+{ runCommand }:
+{
+  a = runCommand "a" { } "echo a > $out";
+  b = runCommand "b" { } "echo b > $out";
+}
+```
+
+```
+$ nix-build -E 'with import <nixpkgs> {}; (callPackages ./callpackages.nix { }).a.override { }'
+this derivation will be built:
+  /nix/store/4sjzxijjfamjqgr8237lr638b8qkabnk-a.drv
+building '/nix/store/4sjzxijjfamjqgr8237lr638b8qkabnk-a.drv'...
+/nix/store/4n3w8mkswwpfa1vvx3012xbaqskflg2z-a
+$ nix-build -E 'with import <nixpkgs> {}; (callPackages ./callpackages.nix { }).a.override { runCommand = runCommandCC; }'
+this derivation will be built:
+  /nix/store/qj9hg9qiahggi4yk6qsh4wv33jl33f36-a.drv
+building '/nix/store/qj9hg9qiahggi4yk6qsh4wv33jl33f36-a.drv'...
+/nix/store/50llkafby4vci46qda0xlva24mlghwr0-a
+```
+
+As you can see, 2 different paths have been produced, due to the fact that we
+replaced the `runCommand`.
```

Comment:
```suggestion
In the Nix package manager, each derivation is uniquely defined, and each corresponding build result produces a unique path in the Nix store, distinguished by a cryptographic hash.

As you can see, two different store paths have been produced for the two different overrides, due to the fact that we set the `runCommand` argument to the value of `runCommandCC` in the second verison.
```

Reply:
Removed in 6491b7cbc30bbff8a36f3e2e809878889b80696f